x-pack/plugins/event_log/server/es/cluster_client_adapter.test.ts

```diff
@@ -195,3 +196,138 @@ describe('createIndex', () => {
     await clusterClientAdapter.createIndex('foo');
   });
 });
+
+describe('queryEventsBySavedObject', () => {
+  test('should call cluster with proper arguments', async () => {
+    clusterClient.callAsInternalUser.mockResolvedValue({
+      hits: {
+        hits: [],
+      },
+    });
+    await clusterClientAdapter.queryEventsBySavedObject(
+      'index-name',
+      'saved-object-type',
+      'saved-object-id',
+      { page: 10, per_page: 10, start: undefined, end: undefined }
+    );
```

Comment:
There seems to be a problem with config schema that forces maybe fields to be explicitly marked as undefined.
I'm looking at replacing our use of schema with fp-ts which is what other solutions are migrating to (infact, we now use it in alerting for our SerDe) which should solve this problem.

Reply:
Ya, it's a pain, which is why I added the DeepWritable/DeepPartial bits for the event interface itself, here: https://github.com/elastic/kibana/blob/cf81ef64cbcad5e8e7e4f8d226b20ca8725e8d95/x-pack/plugins/event_log/generated/schemas.ts#L16-L27

There's still something not quite right about that, IIRC, like the top-level is also optional, so when used, even when declaring a var of that type, `undefined` is a possible value so kind of messes with the typing in a different way.  Typing!

I'm a little worried about fp-ts in that the validation messages aren't going to be as "nice" as config-schema's, but for cases where we don't really need to generate super-nice validation messages (like on IEvent itself), doesn't matter, so hopefully fp-ts will be a nice replacement.  For http P/Q/B and kibana config, I'm a little more worried.

Reply:
BTW, should have mentioned, I'm sure we have other tests - eg, for actions - that pass explicit `undefined` values for the same reason - I think that's fine - we're not expecting customers to be using these types explicitly (at least not yet).  We'll fix 'em all when we get a better schema story ...

Reply:
sigh, I went back and forth and ending up reverting the fp-ts stuff... it requires a lot of boiler plate too for things like default values.
It's great for SerDe, but not great for other things.

I'll try and find a typescript solution in the future... for now using a method like yours :/

Reply:
sigh, I went back and forth and ending up reverting the fp-ts stuff... it requires a lot of boiler plate too for things like default values.
It's great for SerDe, but not great for other things.

I'll try and find a typescript solution in the future... for now using a method like yours :/